agent/src/main/java/com/microsoft/applicationinsights/agent/internal/coresync/impl/ImplementationsCoordinator.java

```diff
@@ -0,0 +1,171 @@
+    public void onException(String className, String methodName, Throwable throwable) {
+        try {
+            AgentNotificationsHandler implementation = getImplementation();
+            if (implementation != null) {
+                implementation.onException(className, methodName, throwable);
+            }
+        } catch (Throwable t) {
+        }
+    }
+
+    @Override
+    public void onMethodEnterURL(String name, URL url) {
+        try {
+            AgentNotificationsHandler implementation = getImplementation();
+            if (implementation != null) {
+                implementation.onMethodEnterURL(name, url);
+            }
+        } catch (Throwable t) {
+        }
+    }
+
+    @Override
+    public void onMethodEnterSqlStatement(String name, Statement statement, String sqlStatement) {
+        try {
+            AgentNotificationsHandler implementation = getImplementation();
+            if (implementation != null) {
+                if (statement != null && sqlStatement != null && sqlStatement.length() == 0) {
+                    sqlStatement = statement.toString();
+                }
+                implementation.onMethodEnterSqlStatement(name, statement, sqlStatement);
+            }
+        } catch (Throwable t) {
+        }
+    }
+
+    @Override
+    public void onDefaultMethodEnter(String name) {
+        try {
+            AgentNotificationsHandler implementation = getImplementation();
+            if (implementation != null) {
+                implementation.onDefaultMethodEnter(name);
+            }
+        } catch (Throwable t) {
+        }
+    }
+
+    @Override
+    public void onMethodFinish(String name, Throwable throwable) {
+        try {
+            AgentNotificationsHandler implementation = getImplementation();
+            if (implementation != null) {
+                implementation.onMethodFinish(name, throwable);
+            }
+        } catch (Throwable t) {
+        }
+    }
+
+    @Override
+    public void onMethodFinish(String name) {
+        try {
+            AgentNotificationsHandler implementation = getImplementation();
+            if (implementation != null) {
+                implementation.onMethodFinish(name);
+            }
+        } catch (Throwable t) {
+        }
+    }
+
+    @Override
+    public String getName() {
+        return null;
+    }
+
+    /**
+     * The data we expect to have for every thread
+     */
+    public static class RegistrationData {
+        public final ClassLoader classLoader;
+        public final AgentNotificationsHandler handler;
+        public final String key;
+
+        public RegistrationData(ClassLoader classLoader, AgentNotificationsHandler handler, String key) {
+            this.classLoader = classLoader;
+            this.handler = handler;
+            this.key = key;
+        }
+    }
+
+    private static ConcurrentHashMap<String, RegistrationData> implementations = new ConcurrentHashMap<String, RegistrationData>();
```

Comment:
Please locate all private fields either before the methods or after these, but not in the middle